Fix Garmin Upload Errors

louisjennings opened the pull request:
Fixes issues #59 , #68 and #70.
The changes were primarily discussed in #68.

Based on branch in PR #75 (as far as I can tell, GitHub doesn't support dependent PRs so all changes are included in this PR as well).

There is currently a known issue, which has caused me to mark this as a draft:

> If an OAuth token becomes invalid or is revoked before expiry, the user will be unable to upload until the token expiry date is reached or the application data is cleared. The OAuth1 token has a validity of 1 year, so neither option is desirable.

I encourage discussion on the best way to solve this. The options I see are:
- Add option in the settings to invalidate tokens (this is the most reliable option but may not be great for UX, though a prompt could be added)
- Automatically invalidate the tokens after a given number of failed login attempts (unrelated errors such as Garmin outages or network faults may invalidate tokens, requiring reauthentication, causing the user to have to re-enter MFA codes. If more than one failed attempt is required to invalidate the tokens, the reset process will be unintuitive)
- Automatically invalidate the tokens only when there is an authentication failure (this is the most user friendly option but may fail if Garmin change how they notify authentication failure or if different types of token invalidation result in different authentication failure messages)

The best solution is probably to combine the first and last options, invalidating the relevant token on a 401 error whilst also allowing the user to clear tokens in case they aren't automatically cleared.

Syntax753 committed:
Initial attempt at fetching oAuth1 and oAuth2 tokens

Syntax753 committed:
Have looked at Scribe but that relies on proper oAuth worflow (Twitter/LinkedIn etc).

signpost seems to be a nice simple implementation. However, it
expects apache HttpRequests when signing:

java.lang.IllegalArgumentException: This consumer expects requests of type org.apache.http.HttpRequest

Will see if cz.msebera.android has helper classes for doing the conversion. Cannot use directly http commons httpclient since Android requires async/non-main thread httpclient calls. cz.msebera.android takes care of that behind the scenes

Syntax753 committed:
oauth values appended to the URI but getting 401 on line 173

Syntax753 committed:
Managed to sign the url with oauth values appended to the URI but getting 401 on line 173 but this suggests the ticket is invalid.

Going via the browser and I get wrong nonce as expected. So the oauth signing does *appear* successful. So need to investigate the invalid ticket number error

louisjennings committed:
Add MVP support for Garmin OAuth
Fix OAuth logic
Update URLs and requests to match garth client
Remove unnecessary redirect code (which prevents OAuth from working)

louisjennings committed:
Add support for MFA codes
Update HTTPClient to non-deprecated version
Add context for some HTTP requests (update the rest)

louisjennings committed:
Clean up code
Extract URI creation to a method
Add members for mfaToken
Add simple class for OAuth2 tokens
Remove unneeded regex patterns
Remove commented out code
Change OAuth token flow (update class members in place)

louisjennings committed:
Add func to save Oauth1 data to SharedPreferences

louisjennings committed:
Add save/load of oauth2 token
Parse remaining oauth2 token members.
Add unused function to load oauth1 token.

louisjennings committed:
Fix issues arising from target api version change

louisjennings committed:
Merge branch 'kochka:master' into update-target-sdk-version

louisjennings committed:
Add a check on refresh token existence and currency.

louisjennings committed:
Remove refresh token logic
Garmin refreshes their OAuth2 token using their OAuth1 token

louisjennings committed:
Merge remote-tracking branch 'origin/unable-to-upload-to-garmin-#68' into unable-to-upload-to-garmin-#68

louisjennings committed:
Upgrade Gradle version
Update targetSdkVersion to 35 (ensuring app can be published on Google Play store)
Migrate android.support libraries to androidx
Fix build issues resulting from changes to versions and dependencies

Changes were tested on emulators targeting API21, API30 and API35. Tested on physical device running API29. Logging and plotting were tested. Data export was not tested. No issues noted.

LLM Output below
-----------------

The following changes were made:

-   Updated compileSdkVersion to 34 and targetSdkVersion to 35.
-   Upgraded the Gradle version and its dependencies.
-   Migrated to AndroidX, replacing `android.support.*` with `androidx.*`.
-   Updated `FloatingActionButton` to use `com.google.android.material.floatingactionbutton.FloatingActionButton`.
-   Updated `Toolbar` to `androidx.appcompat.widget.Toolbar`.
-   Replaced deprecated `getResources().getDrawable()` with `ResourcesCompat.getDrawable()`.
-   Replaced `jcenter` with `mavenCentral`.
-   Updated the fitness API client to use constants from the web archive that were previously in `Scopes`.
-   Added `android.useAndroidX=true` and `android.enableJetifier=true` to gradle.properties.
-   Set `namespace` in build.gradle.
-   Updated `minSdkVersion` in `viewflow/build.gradle`.
- Updated `compileOptions`.
- Update build tools to 35.0.0.
- Fix `appcompat` and material library versions.
- Add baseline_arrow_back_24 icon.
- Updated `httpcore` and `httpmime` dependencies.
- Update kotlin constraints.
- Change excludes to `resources.excludes`.
- Update distribution URL in `gradle-wrapper.properties`.
- Update viewflow dependency
- Updated Google Fit API Client.

louisjennings committed:
Merge branch 'master' into unable-to-upload-to-garmin-#68

# Conflicts:
#	app/build.gradle
#	app/src/main/java/org/kochka/android/weightlogger/tools/GarminConnect.java
#	build.gradle
#	gradle/wrapper/gradle-wrapper.properties

louisjennings committed:
Improve regex for finding CSRF token

louisjennings committed:
Add refresh of OAuth2 token from OAuth1 token, bypassing the need for repeat MFA code entries.

Automatically generated summary below
-------

- Refactor OAuth1 data loading from SharedPreferences.
- Add saving OAuth1 data to SharedPreferences.
- Use the retrieved OAuth1 token to perform an OAuth2 exchange.
- Move OAuth1 loading logic to use shared class variable.
- Update OAuth2 loading to use shared class variable.
- Update OAuth2 exchange to use shared class variable.
- Add initial token loading logic to signin flow.
- Add OAuth1 token saving logic to signin flow.
- Update OAuth2 request to not pass referrer.
- Fix oauth2 and oauth1 token class names.

louisjennings committed:
Add check for OAuth1 expiration before loading it.

Add todo for allowing user to clear the OAuth tokens (in case they become invalid).

louisjennings committed:
Add ability to clear saved Garmin OAuth tokens

Add English language/default strings for the token clearing UI

louisjennings committed:
Re-add saving of OAuth2 token

louisjennings committed:
Add feature: automatically clear relevant OAuth token when HTTP status is 401.

- Changed the error message for failed Garmin Connect login.
- Implemented token clearing when encountering unauthorized errors during login and upload.
- Added functionality to clear OAuth1 and OAuth2 tokens from SharedPreferences.
- Removed now incorrect string translations for failed Garmin Connect login.

louisjennings committed:
Update app version to 2.5.0-beta.1

louisjennings commented:
Added ways to remove auth tokens. They now automatically clear if there is a 401 error during token exchange or during upload. The user can also manually clear them from the settings menu.

Changelog from prior PR comment can be found here:
https://github.com/louisjennings/WeightLogger/releases/tag/v2.5.0-beta.1

Now that the major known issue is resolved, the PR can move out of draft.

louisjennings committed:
Fix bug in OAuth token exchange

The exchange request was inadvertently being sent twice, meaning that the second one failed. This was not picked up in prior testing

louisjennings commented:
The storage issue isn't one that I had anticipated. Storage access is working fine on my parimary physical device running API level 34. I've tried on newer and older emulators and it seems like the problem is with file.mkdirs() on the following line: https://github.com/louisjennings/WeightLogger/blob/8231af43abf076dcf10ca8b8c714c44d90fa4407/app/src/main/java/org/kochka/android/weightlogger/tools/Export.java#L168. The actual error occurs later on where an exception is thrown because the directory doesn't exist. I imagine that the path already exists on my device (maybe from testing before I upgraded the target SDK version), so the export doesn't fail.

I'm not an Android dev either but might be able to work it out. Which lines did you comment out to fix the issue?

Which version of the app did you run? Did you build from the latest commits on my branch or install one of the prebuilt apks? I made a mistake in the beta.1 release, which caused the OAuth exchange to fail. I was so busy testing invalid tokens that I forgot to test a normal auth flow... That was fixed in beta.2, which I released about an hour before your comment. I've tested beta.2 on a couple of devices and have been able to get uploads to work both with and without saved auth tokens. If you have the time, would you be able to try identifying where it's failing? I've been stepping through the signin function in a debugger when troubleshooting signin issues.

The latest tagged release at the time of posting is: https://github.com/louisjennings/WeightLogger/releases/tag/v2.5.0-beta.2

louisjennings committed:
Update version number in advance of future changes.

kochka commented:
Concerning the storage issue, it fail in ```checkPermission``` as ```requestPermissions``` seems to do nothing on my api34 physical device and api35 virtual one.
WRITE_EXTERNAL_STORAGE seems to have been removed in api33.
Need to figure out what should be done, I've bypassed it for the moment :

```java
    if (permission.equals(WRITE_EXTERNAL_STORAGE) && Build.VERSION.SDK_INT > 32) {
      resultListener.onPermissionGranted(permission);
      return;
    }
```

Concerning the login issue, yes I was using your last commit.
Inside ```getOauth1FromResponse```, as it is commented, I've got neither the ```mfaToken```, neither the ```mfaExpirationTimestampString```, witch cause date parsing to fail.
Seems theses data are never used ? I've just omit them and upload is working 👍

louisjennings committed:
Fix bug where exception is thrown when MFA is not enabled on Garmin account.

OAuth1 token responses for accounts without MFA do not contain mfa_expiration_timestamp. Previously, when there was no timestamp, the object would be created with a null timestamp. When adding date parsing, invalid values were not considered. A check for null timestamps has been added. If no expiration is given, token is assumed not to expire.

louisjennings commented:
My latest commit should have fixed the issue where the upload fails if MFA tokens aren't present. I've added a check for `null` `mfa_expiration_timestamp` and set the expiration to the max possible timestamp (indicating that it never expires) if there is no expiration present. As MFA is enabled on my account, I can't check whether there is a differently named parameter in the response. If there is and it isn't being handled, the worst that should happen is that the token gets invalidated after a failed token exchange once it expires

I've tested this change in the debugger by removing `mfaToken` and `mfaExpirationTimestampString` the HTTP response string before it is parsed and both the initial upload and subsequent uploads using saved tokens work as expected.

louisjennings commented:
I've taken a look at the storage and it seems like there are a few different issues:

1. `WRITE_EXTERNAL_STORAGE` no longer has an effect after API level 29, causing the permision check to fail. This can be fixed by checking build version and defaulting to success, as you did.
2. Above API level 29 (maybe below as well?), certain directories such as Downloads and Documents can be written to without permissions. Writing to ~/WeightLogger fails above API level 29. `MANAGE_EXTERNAL_STORAGE` would give permissions to the root of the user storage directory but its use is discouraged as it is intended for applications such as file managers and presents a security/privacy risk to the end user. I would suggest removing the option to export to ~/WeightLogger or changing the option to ~/Documents/WeightLogger.
3. App local data folders have been moved they can be accessed at `context.getFilesDir().getAbsolutePath()`. This also works at lower API levels.

louisjennings committed:
Fix issues with storage permissions.

Generated output below
-----------------------

Update export paths and handle storage permissions

-   Updated export directory paths in `arrays.xml` to use `~/Documents/WeightLogger` and `$appdata/org.kochka.android.weightlogger/files`.
-   Modified `Export.java` to correctly access new export paths, including app data and document directories.
-   Added logic to check `WRITE_EXTERNAL_STORAGE` permission only for API versions below 29 in `PermissionHelper.java`.
-   Added `maxSdkVersion` to the `WRITE_EXTERNAL_STORAGE` permission declaration in `AndroidManifest.xml`, indicating it's not required for API 29 and above.
- Added tools namespace in AndroidManifest to ignore warnings about storage access.

louisjennings committed:
Merge branch 'update-target-sdk-version' into unable-to-upload-to-garmin-#68

louisjennings commented:
I've had a go at implementing those storage fixes in my latest commit. I tested backup and restore to all 4 locations on emulated devices running API 21 and API 30. I also ran exports to all 4 locations on both those API versions and the process completed successfully in all cases. I was only able to check that the files were properly exported to the desired directory on the API 30 device as the API 21 device didn't have a file manager.

Hopefully everything works now!

kochka commented:
I've sent a new version to the Play Store, it's currently in review.
Thx for this long-awaited fix 😁
